static/src/javascripts/projects/common/modules/commercial/slice-adverts.js

```diff
@@ -1,117 +1,103 @@
-                prefs        = userPrefs.get('container-states'),
-                addFabricAd  = (config.page.isFront && config.switches.fabricAdverts && detect.isBreakpoint({max : 'phablet'}));
+    var maxAdsToShow = config.page.showMpuInAllContainers ? 999 : 3;
+    var containerSelector = '.fc-container';
+    var sliceSelector = '.js-fc-slice-mpu-candidate';
+    var containerGap = 1;
 
-            // insert fabric advert at first container in lieu of a top slot
-            if (addFabricAd) {
-                fabricAdSlot = bonzo(createAdSlot('fabric', 'container-inline'));
-                fabricAdSlot.addClass('ad-slot--mobile');
-                fabricAdSlot.insertAfter(containers[0]);
-            }
+    return {
+        init: init
+    };
 
-            // skip the initial containers if we've already added an advert to the first
-            index+= addFabricAd ? (1 + containerGap) : 0;
+    function init() {
+        if (!commercialFeatures.sliceAdverts) {
+            return false;
+        }
 
-            // pull out ad slices which have at least x containers between them
-            while (index < containers.length) {
-                container    = containers[index];
-                containerId  = bonzo(container).data('id');
-                $adSlice     = $(opts.sliceSelector, container);
-                // don't display ad in the first container on the network fronts
-                isFrontFirst = contains(['uk', 'us', 'au'], config.page.pageId) && index === 0;
+        var adSlots;
+        var prefs          = userPrefs.get('container-states') || {};
+        var isMobile       = detect.getBreakpoint() === 'mobile';
+        var isNetworkFront = ['uk', 'us', 'au'].indexOf(config.page.pageId) !== -1;
+        var hasFabricAd    = (config.page.isFront && config.switches.fabricAdverts && detect.isBreakpoint({max : 'phablet'}));
+        // We must keep a small bit of state in the filtering logic
+        var lastIndex      = -1;
 
+        adSlots = qwery(containerSelector)
+            // get all ad slices
+            .map(function (container) {
+                return {
+                    container: container,
+                    adSlice: container.querySelector(sliceSelector)
+                };
+            })
+            // pull out closed, empty (no slice) or first on front containers,
+            // keeping containers only if they are $containerGap nodes apart
+            .filter(function (item, index) {
                 if (config.page.showMpuInAllContainers) {
-                    adSlices.push($adSlice.first());
-                    index++;
-                } else {
-                    if ($adSlice.length && !isFrontFirst && (!prefs || prefs[containerId] !== 'closed')) {
-                        adSlices.push($adSlice.first());
-                        index += (containerGap + 1);
-                    } else {
-                        index++;
-                    }
+                    return true;
```

Comment:
The original code, in this case, only pushes the slices it finds in each container. This pushes all of the containers themselves. If the switch is on and some of the containers lack ad candidates, could the call to `anchor.nextSibling` below occasionally throw an exception?